### PR TITLE
Allowed Filter param to be none everywhere in erlcloud_ec2

### PR DESCRIPTION
**Pull request by SergeyOvchinnik**
Reverted changes restricting Filter param from being none and allowed it in all filterable describe functions

### CODE REVIEW

File: src/erlcloud_ec2.erl [hunk 1]
```diff
@@ -3188,26 +3196,32 @@ describe_tags(Filters, Config)
...
-                I = integer_to_list(Index),
-                Key = "Filter."++I++".Name",
-                Prefix = "Filter."++I++".Value.",
-                {value_list_params(Values, Prefix) ++ [{Key, filter_name(Name)} | Acc], Index + 1}
-            end, {[], 1}, Filters),
+    when is_list(Filters) orelse Filters =:= none,
+         is_integer(MaxResults), MaxResults >= ?TAGS_MR_MIN, MaxResults =< ?TAGS_MR_MAX,
+         is_list(NextToken) orelse NextToken =:= undefined,
+         is_record(Config, aws_config) ->
+    FilterParams = case Filters of  %%TODO replace with list_to_ec2_filter(Filters) ?
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: just do it

File: src/erlcloud_ec2.erl [hunk 2]
```diff
@@ -3162,24 +3164,30 @@ describe_tags() ->
...
-        lists:foldl(
-            fun({Name, Values}, {Acc, Index}) ->
-                I = integer_to_list(Index),
-                Key = "Filter."++I++".Name",
-                Prefix = "Filter."++I++".Value.",
-                {value_list_params(Values, Prefix) ++ [{Key, filter_name(Name)} | Acc], Index + 1}
-            end, {[], 1}, Filters),
-    case ec2_query(Config, "DescribeTags", Params, ?NEW_API_VERSION) of
+    when is_list(Filters) orelse Filters =:= none, is_record(Config, aws_config) ->
+    FilterParams = case Filters of  %%TODO replace with list_to_ec2_filter(Filters) ?
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: just do it

File: src/erlcloud_ec2.erl (hunk 1 above)
Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl]: Can't do that, describe_tags accepts filter names as atoms and list_to_ec2_filter doesn't handle them well

Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl]: Actually, I might be able to rewrite list_to_ec2_filter a bit

**Comment by kkuzmin** [comment]
Please, propagate the fix to erlcloud/erlcloud

File: src/erlcloud_ec2.erl [hunk 3]
```diff
@@ -3375,6 +3347,10 @@ list_to_ec2_filter(List) ->
 
 list_to_ec2_filter([], _Count, Res) ->
     Res;
+list_to_ec2_filter([{N, V}|T], Count, Res)
+    when is_atom(N) ->
+    NewName = [case Char of 95 -> 45; _ -> Char end || Char <- atom_to_list(N)],
```
Comment (kkuzmin) [inline src/erlcloud_ec2.erl]: Please use `$_` and `$-` instead of char codes to make it more readable.

File: src/erlcloud_ec2.erl [hunk 4]
```diff
@@ -3375,6 +3347,10 @@ list_to_ec2_filter(List) ->
 
 list_to_ec2_filter([], _Count, Res) ->
     Res;
+list_to_ec2_filter([{N, V}|T], Count, Res)
+    when is_atom(N) ->
+    NewName = [case Char of $_ -> $-; _ -> Char end || Char <- atom_to_list(N)],
```
Comment (motobob) [inline src/erlcloud_ec2.erl:3349]: black magic detected. this might hit hard in future. 
what is the motivation for this change

**Comment by motobob** [comment]
https://github.com/alertlogic/erlcloud/pull/136/files#diff-1e596cd7a553dd4a4965a3193b91eb37R208 should be updated

**Comment by motobob** [comment]
frankly, how about removing `when is_list(Filter) orelse Filter =:= none ` everywhere?

File: src/erlcloud_ec2.erl (hunk 4 above)
Comment (SergeyOvchinnik) [reply src/erlcloud_ec2.erl:3349]: This converts atoms 'resource_type' to strings "resource-type". The motivation for this change was to replace the weird Filter handling in describe_tags with just one list_to_ec2_filter call. describe_tags used to use atoms as filter names unlike other describe functions which used strings. The only way to implement filter handling in describe_tags using list_to_ec2_filter and ensure backward compatibility was to allow list_to_ec2_filter to accept atoms as filter names, which is exactly what this clause does. If filter name is provided as atom, it converts it to appropriate string and calls itself with the string filter name.

Comment (motobob) [reply src/erlcloud_ec2.erl:3349]: okay

**Comment by motobob** [comment]
👍 but i would suggest we need to clean up guards madness

**Comment by motobob** [comment]
@kkuzmin your call

File: src/erlcloud_ec2.erl [hunk 5]
```diff
@@ -205,7 +205,7 @@
 -define(RESERVED_INSTANCES_OFFERINGS_MR_MIN, 5).
 -define(RESERVED_INSTANCES_OFFERINGS_MR_MAX, 1000).
 
--type filter_list() :: [{string(),[string()]}].
+-type filter_list() :: [{string() | atom(),[string()]}] | none().
```
Comment (kkuzmin) [inline src/erlcloud_ec2.erl]: I suspect that `none` should be an atom.

**Comment by kkuzmin** [comment]
+1

**Comment by motobob** [comment]
@SergeyOvchinnik please make a clone PR to upstream and no more PR accepted here

**Pull request merged**